src/providers/idp/idp_auth.c

```diff
@@ -0,0 +1,543 @@
+              sss_authtok_get_type(pd->authtok),
+              sss_authtok_type_to_str(sss_authtok_get_type(pd->authtok)));
+        goto done;
+    }
+
+    ret = sss_authtok_get_oauth2(pd->authtok, &user_code, &user_code_len);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to extract user code.\n");
+        goto done;
+    }
+
+    open_req = sss_ptr_hash_lookup(idp_auth_ctx->open_request_table,
+                                   user_code, struct idp_open_req_data);
+    if (open_req == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to retrieve stored request data.\n");
+        goto done;
+    }
+    if (open_req->device_code_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Missing device code data.\n");
+        goto done;
+    }
+
+    send_data = talloc_asprintf(mem_ctx, "%s\n%s",
+                                dp_opt_get_cstring(idp_auth_ctx->idp_options,
+                                                   IDP_CLIENT_SECRET),
+                                open_req->device_code_data);
+    if (send_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to generate auth data.\n");
+        goto done;
+    }
+
+done:
+    talloc_free(open_req);
+
+    return send_data;
+}
+
+static errno_t create_auth_send_buffer(TALLOC_CTX *mem_ctx,
+                                       struct idp_auth_ctx *idp_auth_ctx,
+                                       struct pam_data *pd,
+                                       struct io_buffer **io_buf)
+{
+    struct io_buffer *buf = NULL;
+    const char *send_data;
+    int ret;
+
+    buf = talloc_zero(mem_ctx, struct io_buffer);
+    if (buf == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc failed.\n");
+        return ENOMEM;
+    }
+
+    switch (pd->cmd) {
+    case SSS_PAM_PREAUTH:
+        send_data = dp_opt_get_cstring(idp_auth_ctx->idp_options,
+                                       IDP_CLIENT_SECRET);
+        if (send_data == NULL || *send_data == '\0') {
+            ret = EOK;
+            goto done;
+        }
+        break;
+    case SSS_PAM_AUTHENTICATE:
+        send_data = get_stored_request_data(buf, idp_auth_ctx, pd);
+        if (send_data == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "Failed to get stored device code data.\n");
+            ret = ENOENT;
+            goto done;
+        }
+        break;
+    default:
+        DEBUG(SSSDBG_OP_FAILURE, "Unsupported pam task [%d][%s].\n",
+                                 pd->cmd, sss_cmd2str(pd->cmd));
+        ret = EINVAL;
+        goto done;
+    }
+
+    buf->size = strlen(send_data);
+    buf->data = talloc_size(buf, buf->size);
+    if (buf->data == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_size failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    safealign_memcpy(buf->data, send_data, buf->size, NULL);
+
+    ret = EOK;
+
+done:
+    if (ret == EOK) {
+        *io_buf = buf;
+    } else {
+        talloc_free(buf);
+    }
+
+    return ret;
```

Comment:
Mem allocated in `get_stored_request_data()` is attached to `buf`, so strictly speaking doesn't leak here...

Maybe simply `buf->data = send_data` then? This should work for `SSS_PAM_PREAUTH` case as well if not modified later.

Reply:
Maybe a bad idea.